apps/rush-lib/src/logic/PackageChangeAnalyzer.ts

```diff
@@ -71,17 +76,30 @@ export class PackageChangeAnalyzer {
     return projectState;
   }
 
-  private _getData(): Map<string, Map<string, string>> | undefined {
+  private async _getData(terminal: Terminal): Promise<Map<string, Map<string, string>> | undefined> {
     const repoDeps: Map<string, string> | undefined = this._getRepoDeps();
     if (!repoDeps) {
       return undefined;
     }
 
     const projectHashDeps: Map<string, Map<string, string>> = new Map<string, Map<string, string>>();
 
-    // pre-populate the map with the projects from the config
+    const ignorePatternsByProject: Map<string, IMinimatch[]> = new Map<string, IMinimatch[]>();
+
+    // pre-populate the map with the projects from the config, and load all project configurations up-front
     for (const project of this._rushConfiguration.projects) {
       projectHashDeps.set(project.packageName, new Map<string, string>());
+
+      const projectConfiguration:
+        | RushProjectConfiguration
+        | undefined = await RushProjectConfiguration.tryLoadForProjectAsync(project, undefined, terminal);
```

Comment:
We'd want to do an async parallel invocation if we are waiting for file reads in this loop

Reply:
This function is very much critical path for all rush commands that do any work

Reply:
Makes sense! I was hunting around the codebase and I just might not have found it yet -- is there a generic bottleneck / rate limiting / parallelism-limited async map somewhere in rush-lib that I can use?

(I'm pretty sure dropping a `Promise.all(map)` bomb is actually slower for 200+ files, so I guess I'd feel better if we could limit it to a parallelism of 16 or something.)

Reply:
>  I was hunting around the codebase and I just might not have found it yet -- is there a generic bottleneck / rate limiting / parallelism-limited async map somewhere in rush-lib that I can use?

@elliot-nelson Weirdly no, I can't seem to find one.  (Which is weird -- I'm sure someone encountered this before!)  The idiomatic pattern would be an API something like:

```js
import { Async } from '@rushstack/node-core-library';

// (or whatever the right API design is)
Async.throttle(3,
  [
    () => doAsyncThing(),
    () => doAsyncThing(),
    () => doAsyncThing(),
    () => doAsyncThing(),
    () => doAsyncThing(),
    () => doAsyncThing(),
  ] 
);
```

If you want to recommend a high quality throttling NPM package, we could use this PR to add a `@beta` API  in `@rushstack/node-core-library` that wraps it.

Reply:
I put up a PR with an alternate pretty quick option -- https://github.com/microsoft/rushstack/pull/2665.